crates/ruff_python_semantic/src/red_knot/semantic_index.rs

```diff
@@ -0,0 +1,652 @@
+
+        assert_eq!(names(&root_table), vec!["foo"]);
+        let foo = root_table.symbol_by_name("foo").unwrap();
+
+        assert_eq!(foo.definitions().len(), 1);
+    }
+
+    #[test]
+    fn import_sub() {
+        let TestCase { db, file } = test_case("import foo.bar");
+        let root_table = symbol_table(&db, root_scope(&db, file));
+
+        assert_eq!(names(&root_table), vec!["foo"]);
+    }
+
+    #[test]
+    fn import_as() {
+        let TestCase { db, file } = test_case("import foo.bar as baz");
+        let root_table = symbol_table(&db, root_scope(&db, file));
+
+        assert_eq!(names(&root_table), vec!["baz"]);
+    }
+
+    #[test]
+    fn import_from() {
+        let TestCase { db, file } = test_case("from bar import foo");
+        let root_table = symbol_table(&db, root_scope(&db, file));
+
+        assert_eq!(names(&root_table), vec!["foo"]);
+        assert_eq!(
+            root_table
+                .symbol_by_name("foo")
+                .unwrap()
+                .definitions()
+                .len(),
+            1
+        );
+        assert!(
+            root_table
+                .symbol_by_name("foo")
+                .is_some_and(|symbol| { symbol.is_defined() || !symbol.is_used() }),
+            "symbols that are defined get the defined flag"
+        );
+    }
+
+    #[test]
+    fn assign() {
+        let TestCase { db, file } = test_case("x = foo");
+        let root_table = symbol_table(&db, root_scope(&db, file));
+
+        assert_eq!(names(&root_table), vec!["foo", "x"]);
+        assert_eq!(
+            root_table.symbol_by_name("x").unwrap().definitions().len(),
+            1
+        );
+        assert!(
+            root_table
+                .symbol_by_name("foo")
+                .is_some_and(|symbol| { !symbol.is_defined() && symbol.is_used() }),
+            "a symbol used but not defined in a scope should have only the used flag"
+        );
+    }
+
+    #[test]
+    fn class_scope() {
+        let TestCase { db, file } = test_case(
+            "
+class C:
+    x = 1
+y = 2
+",
+        );
+        let root_table = symbol_table(&db, root_scope(&db, file));
+
+        assert_eq!(names(&root_table), vec!["C", "y"]);
+
+        let index = semantic_index(&db, file);
+
+        let scopes: Vec<_> = index.child_scopes(ScopeId::root()).collect();
+        assert_eq!(scopes.len(), 1);
+
+        let (class_scope_id, class_scope) = scopes[0];
+        assert_eq!(class_scope.kind(), ScopeKind::Class);
+        assert_eq!(class_scope.name(), "C");
+
+        let class_table = index.symbol_table(class_scope_id);
+        assert_eq!(names(&class_table), vec!["x"]);
+        assert_eq!(
+            class_table.symbol_by_name("x").unwrap().definitions().len(),
+            1
+        );
+    }
+
+    #[test]
+    fn function_scope() {
+        let TestCase { db, file } = test_case(
+            "
+def func():
+    x = 1
+y = 2
+",
```

Comment:
I would prefer to use `textwrap::dedent` in `test_case` function and avoid this ugly (lack of) indentation.

Reply:
One reason I dislike the use of `dedent` is because it makes it very difficult to reason about text offsets because they no longer map to the source you see in the test.

Reply:
Ah, hm. I do think the indented version is significantly more readable (it doesn't disrupt the visual indentation of the overall block structure), but I do see the point about offsets. I'm not totally convinced by it, though.

I'm not sure how often I would ever manually count character offsets in the text on screen, but even if I did, it would feel pretty natural to skip all those leading spaces rather than try to count them.

And if the issue is programmatically indexing into the text within the test, that's also not a problem if the version of the text that accessible to the test is the already-dedented one. For instance, in these tests, if the dedent happens in `test_case`, and the returned `TestCase` exposes the dedented text, it's not an issue.